File: Framework/LanguageSupport/cpp/MMICPP/Adapter/MMIScene.cpp
#include "MMIScene.h"
#include "Extensions/MVector3Extensions.h"
#include "Extensions/MQuaternionExtensions.h"
#include "boost/exception/diagnostic_information.hpp"
#include "Utils/Logger.h"
#include <iostream>
#include "Extensions/MBoolResponseExtensions.h"

MMIScene::MMIScene():frameID{0},historyBufferSize{20}
{
}

void MMIScene::GetSceneObjects(std::vector<MSceneObject>& _return)
{
	for(const pair <string, MSceneObject> &ob: this->sceneObjectsById)
	{
		_return.emplace_back(ob.second);
	}
}

shared_ptr<vector<MSceneObject>> MMIScene::GetSceneObjects()
{
	auto _return =make_shared<vector< MSceneObject>>();
	this->GetSceneObjects(*_return);
	return _return;
}

void MMIScene::GetSceneObjectByID(MSceneObject & _return, const std::string & id)
{
	auto iter = this->sceneObjectsById.find(id);
	if (iter != sceneObjectsById.end())
		_return = iter->second;
}

shared_ptr<MSceneObject> MMIScene::GetSceneObjectByID(const string & id)
{
	auto _return = make_shared<MSceneObject>();
	this->GetSceneObjectByID(*_return, id);
	return _return;
}

void MMIScene::GetSceneObjectByName(MSceneObject & _return, const std::string & name)
{
	auto iter = nameIdMappingSceneObjects.find(name);
	if (iter != nameIdMappingSceneObjects.end())
		this->GetSceneObjectByID(_return,iter->second[0]);
}

shared_ptr<MSceneObject> MMIScene::GetSceneObjectByName(const string & name)
{
	auto _return = make_shared<MSceneObject>();
	this->GetSceneObjectByName(*_return, name);
	return _return;
}

void MMIScene::GetSceneObjectsInRange(std::vector<MSceneObject>& _return, const::MMIStandard::MVector3 & position, const double range)
{
	vector<MSceneObject> objects;
	this->GetSceneObjects(objects);
	for (const MSceneObject &ob : objects)
	{
		if (MVector3Extensions::EuclideanDistance(ob.Transform.Position, position) <= range)
		{
			_return.emplace_back(ob);
		}	
	}
}

shared_ptr<vector<MSceneObject>> MMIScene::GetSceneObjectsInRange(const MVector3 & position, const double range)
{
	auto _return = make_shared <vector<MSceneObject>>();
	this->GetSceneObjectsInRange(*_return, position, range);
	return _return;
}

void MMIScene::GetColliders(std::vector<MCollider>& _return)
{
	vector<MSceneObject> sceneObjects;
	this->GetSceneObjects(sceneObjects);

	for (const MSceneObject &ob : sceneObjects)
	{
		_return.emplace_back(ob.Collider);
	}
}

shared_ptr<vector<MCollider>> MMIScene::GetColliders()
{
	auto _return = make_shared<vector<MCollider>>();
	this->GetColliders(*_return);
	return _return;
}

void MMIScene::GetColliderById(MCollider & _return, const std::string & id)
{
	MSceneObject sceneObject;
	this->GetSceneObjectByID(sceneObject, id);
	_return = sceneObject.Collider;
}

shared_ptr<MCollider> MMIScene::GetColliderById(const string & id)
{
	auto _return = make_shared<MCollider>();
	this->GetColliderById(*_return,id);
	return _return;
}

void MMIScene::GetCollidersInRange(std::vector<MCollider>& _return, const::MMIStandard::MVector3 & position, const double range)
{
	vector<MSceneObject> sceneObjects;
	this->GetSceneObjectsInRange(sceneObjects, position, range);

	for (const MSceneObject &ob : sceneObjects)
	{
		if (ob.__isset.Collider)
		{
			_return.emplace_back(ob.Collider);
		}
	}
}

shared_ptr<vector<MCollider>> MMIScene::GetCollidersInRange(const::MMIStandard::MVector3 & position, const double range)
{
	auto _return = make_shared<vector<MCollider>>();
	this->GetCollidersInRange(*_return,position, range);
	return _return;
}

void MMIScene::GetMeshes(std::vector<MMesh>& _return)
{
	vector<MSceneObject> sceneObjects;
	this->GetSceneObjects(sceneObjects);
	for (const MSceneObject &sceneObject: sceneObjects)
		_return.emplace_back(sceneObject.Mesh);
}

shared_ptr<vector<MMesh>> MMIScene::GetMeshes()
{
	auto _return = make_shared<vector<MMesh>>();
	this->GetMeshes(*_return);
	return _return;
}

void MMIScene::GetMeshByID(MMesh & _return, const std::string & id)
{
	MSceneObject sceneObject;
	this->GetSceneObjectByID(sceneObject, id);
	_return = sceneObject.Mesh;
}

shared_ptr<MMesh> MMIScene::GetMeshByID(const std::string & id)
{
	auto _return = make_shared<MMesh>();
	this->GetMeshByID(*_return, id);
	return _return;
}

void MMIScene::GetTransforms(std::vector<MTransform>& _return)
{
	vector<MSceneObject> sceneObjects;
	this->GetSceneObjects(sceneObjects);
	for (const MSceneObject &sceneObject : sceneObjects)
		_return.emplace_back(sceneObject.Transform);
}

shared_ptr<vector<MTransform>> MMIScene::GetTransforms()
{
	auto _return = make_shared<vector<MTransform>>();
	this->GetTransforms(*_return);
	return _return;
}

void MMIScene::GetTransformByID(MTransform & _return, const std::string & id)
{
	MSceneObject sceneObject;
	this->GetSceneObjectByID(sceneObject, id);
	_return = sceneObject.Transform;
}

shared_ptr<MTransform> MMIScene::GetTransformByID(string & id)
{
	auto _return = make_shared<MTransform>();
	this->GetTransformByID(*_return,id);
	return _return;
}

void MMIScene::GetAvatars(std::vector<MAvatar>& _return)
{
	for(const auto &ob : this->avatarsById)
	{
		_return.emplace_back(ob.second);
	}
}

shared_ptr<vector<MAvatar>> MMIScene::GetAvatars()
{
	auto _return =make_shared<vector<MAvatar>>();
	this->GetAvatars(*_return);
	return _return;
}

void MMIScene::GetAvatarByID(MAvatar & _return, const std::string & id)
{
	auto iter = this->avatarsById.find(id);
	if (iter != avatarsById.end())
		_return = iter->second;
}

shared_ptr<MAvatar> MMIScene::GetAvatarByID(const string & id)
{
	auto _return = make_shared<MAvatar>();
	this->GetAvatarByID(*_return, id);
	return _return;
}

void MMIScene::GetAvatarByName(MAvatar & _return, const std::string & name)
{
	auto iter = nameIdMappingAvatars.find(name);
	if (iter != nameIdMappingAvatars.end())
		this->GetAvatarByID(_return, iter->second[0]);
}

shared_ptr<MAvatar> MMIScene::GetAvatarByName(const string & name)
{
	auto _return = make_shared<MAvatar>();
	this->GetAvatarByName(*_return, name);
	return _return;
}

void MMIScene::GetAvatarsInRange(std::vector<MAvatar>& _return, const::MMIStandard::MVector3 & position, const double distance)
{
	vector<MAvatar> avatars;
	this->GetAvatars(avatars);

	for (MAvatar avatar : avatars)
	{
		MVector3 avatarPosition = MVector3{};
		avatarPosition.__set_X(avatar.PostureValues.PostureData[0]);
		avatarPosition.__set_Y(avatar.PostureValues.PostureData[1]);
		avatarPosition.__set_Z(avatar.PostureValues.PostureData[3]);

		if (MVector3Extensions::EuclideanDistance(avatarPosition, position) <= distance)
		{
			_return.emplace_back(avatar);
		}
	}
}

shared_ptr<vector<MAvatar>> MMIScene::GetAvatarsInRange(const::MMIStandard::MVector3 & position, const double distance)
{
	auto _return = make_shared<vector<MAvatar>>();
	this->GetAvatarsInRange(*_return, position, distance);
	return _return;
}

double MMIScene::GetSimulationTime()
{
	throw std::runtime_error("Function not implemented yet");
}

void MMIScene::GetSceneChanges(MSceneUpdate & _return)
{
	_return = this->sceneUpdate;
}

shared_ptr<MSceneUpdate> MMIScene::GetSceneChanges()
{
	return make_shared<MSceneUpdate>(this->sceneUpdate);
}


//TODO check
void MMIScene::GetFullScene(MSceneUpdate & _return)
{

	this->GetSceneObjects(_return.AddedSceneObjects);
	this->GetAvatars(_return.AddedAvatars);
}

shared_ptr<MSceneUpdate> MMIScene::GetFullScene()
{
	auto _return = make_shared<MSceneUpdate>();
	this->GetFullScene(*_return);
	return _return;
}

void MMIScene::GetNavigationMesh(MNavigationMesh & _return)
{
	throw std::runtime_error("Function not implemented yet");
}

shared_ptr<MNavigationMesh> MMIScene::GetNavigationMesh()
{
	throw std::runtime_error("Function not implemented yet");
}

void MMIScene::Clear()
{
	this->avatarsById.clear();
	this->sceneObjectsById.clear();
	this->nameIdMappingAvatars.clear();
	this->nameIdMappingSceneObjects.clear();
	this->sceneUpdate.~MSceneUpdate();
	this->sceneUpdate =  MSceneUpdate{};
	this->frameID = 0;
	this->sceneHistory.clear();
}

void MMIScene::Apply(MBoolResponse & _return, const MSceneUpdate & sceneUpdate)
{
	_return.__set_Successful(true);
	this->frameID++;
	this->sceneHistory.emplace_front(pair(frameID, sceneUpdate));
	while (this->sceneHistory.size() > (size_t)this->historyBufferSize)
		sceneHistory.pop_back();

	this->sceneUpdate = sceneUpdate;

	if(sceneUpdate.__isset.AddedAvatars)
		this->AddAvatars(_return,sceneUpdate.AddedAvatars);
	
	if (sceneUpdate.__isset.AddedSceneObjects)
		this->AddSceneObjects(_return, sceneUpdate.AddedSceneObjects);

	if (sceneUpdate.__isset.ChangedAvatars)
		this->UpdataAvatars(_return,sceneUpdate.ChangedAvatars);

	if (sceneUpdate.__isset.ChangedSceneObjects)
		this->UpdateSceneObjects(_return,sceneUpdate.ChangedSceneObjects);

	if (sceneUpdate.__isset.RemovedAvatars)
		this->RemoveAvatars(_return,sceneUpdate.RemovedAvatars);

	if (sceneUpdate.__isset.RemovedSceneObjects)
		this->RemoveSceneObjects(_return,sceneUpdate.RemovedSceneObjects);
}


void MMIScene::AddAvatars(MBoolResponse & _return, const vector<MAvatar>& avatars)
{
	
	for (const MAvatar &avatar : avatars)
	{
		if(!this->avatarsById.insert(std::pair<string, MAvatar>(avatar.ID, avatar)).second) // try to insert new avatar return is false if avatar is already in the map
		{
			string message = "Could not add avatar: " + avatar.Name + " is already registered";
			Logger::printLog(L_ERROR, message);
			MBoolResponseExtensions::Update(_return, message, false);
			continue;
		}

		auto iter1 = this->nameIdMappingAvatars.find(avatar.Name);
		if (iter1 != nameIdMappingAvatars.end()) // contains already the avatar
		{
			
			if (!(std::find(iter1->second.begin(), iter1->second.end(), avatar.ID) != iter1->second.end()))
				this->nameIdMappingAvatars[avatar.Name].emplace_back(avatar.ID);
		}
		else
		{
			nameIdMappingAvatars.insert(std::pair < string, vector<string>>(avatar.Name, vector<string>{avatar.ID}));
		}
	}
}

void MMIScene::AddSceneObjects(MBoolResponse & _return, const vector<MSceneObject>& sceneObjects)
{
	for (const MSceneObject &sceneObject: sceneObjects)
	{
		if (!this->sceneObjectsById.insert(std::pair<string, MSceneObject>(sceneObject.ID, sceneObject)).second)
		{
			string message = "Could not add scene object: " + sceneObject.Name + " is already registered";
			Logger::printLog(L_ERROR, message);
			MBoolResponseExtensions::Update(_return, message, false);
			continue;
		}

		auto iter1 = this->nameIdMappingSceneObjects.find(sceneObject.Name);
		if (iter1 != nameIdMappingSceneObjects.end())
		{
			if (!(std::find(iter1->second.begin(), iter1->second.end(), sceneObject.ID) != iter1->second.end()))
				this->nameIdMappingSceneObjects[sceneObject.Name].emplace_back( sceneObject.ID );
		}
		else
		{
			nameIdMappingSceneObjects.insert(std::pair < string, vector<string>>(sceneObject.Name, vector<string>{sceneObject.ID}));
		}
	}
}

void MMIScene::UpdataAvatars(MBoolResponse & _return, const vector<MAvatarUpdate>& avatars)
{
	for (MAvatarUpdate avatarUpdate : avatars)
	{
		auto iter = this->avatarsById.find(avatarUpdate.ID);
		if (iter != avatarsById.end())
		{
			if (avatarUpdate.__isset.Description)
				iter->second.Description = avatarUpdate.Description;

			if (avatarUpdate.__isset.PostureValues)
				iter->second.PostureValues = avatarUpdate.PostureValues;

			if (avatarUpdate.__isset.SceneObjects)
				iter->second.SceneObjects = avatarUpdate.SceneObjects;
		}
		else
		{
			string message = "Could not update avatar : " + avatarUpdate.ID + " not found";
			Logger::printLog(L_ERROR, message);
			MBoolResponseExtensions::Update(_return, message, false);
		}
	}
}

void MMIScene::UpdateSceneObjects(MBoolResponse & _return, const vector<MSceneObjectUpdate>& sceneObjects)
{
	for (MSceneObjectUpdate sceneObjectUpdate : sceneObjects)
	{
		auto iter = this->sceneObjectsById.find(sceneObjectUpdate.ID);
		if (iter != sceneObjectsById.end())
		{
			if (sceneObjectUpdate.__isset.Transform)
			{
				MTransformUpdate transformUpdate = sceneObjectUpdate.Transform;
				try
				{
					if(transformUpdate.__isset.Position)
						MVector3Extensions::ToMVector3(iter->second.Transform.Position, sceneObjectUpdate.Transform.Position);
					
					if(transformUpdate.__isset.Rotation)
						MQuaternionExtensions::ToMQuaternion(iter->second.Transform.Rotation, sceneObjectUpdate.Transform.Rotation);
				}
				catch (...)
				{
					std::clog << boost::current_exception_diagnostic_information() << std::endl;
					_return.__set_Successful(false);
					if (_return.__isset.LogData)
					{
						_return.LogData.emplace_back(boost::current_exception_diagnostic_information());
					}
					else
					{
						_return.__set_LogData(vector<string>{ boost::current_exception_diagnostic_information() });
					}
				}
				if (transformUpdate.__isset.Parent)
					iter->second.Transform.Parent = sceneObjectUpdate.Transform.Parent;
			}

			if (sceneObjectUpdate.__isset.Collider)
				iter->second.Collider=sceneObjectUpdate.Collider;

			if (sceneObjectUpdate.__isset.Mesh)
				iter->second.Mesh = sceneObjectUpdate.Mesh;

			if (sceneObjectUpdate.__isset.PhysicsProperties)
				iter->second.PhysicsProperties = sceneObjectUpdate.PhysicsProperties;
		}
		else
		{
			string message = "Could not update scene object: " + sceneObjectUpdate.ID + " object not found";
			Logger::printLog(L_ERROR, message);
			MBoolResponseExtensions::Update(_return, message, false);
		}
	}
}

void MMIScene::RemoveAvatars(MBoolResponse & _return, const vector<string>& avatarIDs)
{
	MBoolResponse response{};
	response.Successful = true;

	for (string id : avatarIDs)
	{
		auto avatarIter = this->avatarsById.find(id);
		if (avatarIter != avatarsById.end()) //pos avatar in avatarsById
		{
			auto nameIdIter = this->nameIdMappingAvatars.find(avatarIter->second.Name);
			if (nameIdIter != nameIdMappingAvatars.end()) //pos vector<string> of ids
			{
				auto idIter = std::find(nameIdIter->second.begin(), nameIdIter->second.end(), id); //pos string in the ids vector<string>
				if (idIter != nameIdIter->second.end())
					nameIdIter->second.erase(idIter);
			}

			avatarsById.erase(avatarIter);
		}
		else
		{
			string message = "could not remove avatar: " + id + " not found";
			Logger::printLog(L_ERROR, message);
			MBoolResponseExtensions::Update(_return, message, false);
		}
	}

}

void MMIScene::RemoveSceneObjects(MBoolResponse & _return, const vector<string>& sceneObjectIDs)
{
	MBoolResponse response{};
	response.Successful = true;

	for (string id : sceneObjectIDs)
	{
		auto iter = this->sceneObjectsById.find(id);
		if (iter != sceneObjectsById.end())
		{
			auto iter1 = this->nameIdMappingSceneObjects.find(iter->second.Name);
			if (iter1 != nameIdMappingSceneObjects.end())
			{
				auto iter2 = std::find(iter1->second.begin(), iter1->second.end(), id);
				if (iter2 != iter1->second.end())
					iter1->second.erase(iter2);
			}

			sceneObjectsById.erase(iter);
		}
		else
		{
			string message = "could not remove scene object:  " + id + " not found";
			Logger::printLog(L_ERROR, message);
			MBoolResponseExtensions::Update(_return, message, false);
		}
	}
}

File: Framework/LanguageSupport/java/MMIJava/src/Adapter/MMIScene.java
package Adapter;

import Extensions.MQuaternionExtensions;
import Extensions.MVector3Extensions;
import MMIStandard.*;

import java.util.*;
import java.util.stream.Collectors;

public class MMIScene implements MSceneAccess.Iface {
    /*
            Class represents a (hypothetical) scene which can be specifically set up by the developer
    */
    //	Map containing all scene objects structured by the specific id
    private final Map<String, MSceneObject> sceneObjectsByID = new HashMap<>();
    //	Map containing all avatars structured by the specific id
    private final Map<String, MAvatar> avatarsByID = new HashMap<>();
    //	Mapping between the name of a scene object and a unique id
    private final Map<String, List<String>> nameIdMappingSceneObjects = new HashMap<>();
    //	Mapping between the name of a avatar and a unique id
    private final Map<String, List<String>> nameIdMappingAvatars = new HashMap<>();
    //	MSceneUpdate from the previous frame
    private MSceneUpdate sceneUpdate = new MSceneUpdate();
    //	ID of the frame
    private int frameID = 0;
    //	A list  which contains the history of the last n applied scene manipulations
    private Queue<Tuple<Integer, MSceneUpdate>> SceneHistory = new LinkedList<>();
    //	The size of the HistoryBuffer
    private final int HistoryBufferSize = 20;

    //	Clears the whole scene
    public void clear() {
        this.avatarsByID.clear();
        this.sceneObjectsByID.clear();
        this.nameIdMappingAvatars.clear();
        this.nameIdMappingSceneObjects.clear();
        this.SceneHistory.clear();
        this.sceneUpdate = new MSceneUpdate();
        this.frameID = 0;
    }

    //	Returns the scene objects
    @Override
    public List<MSceneObject> GetSceneObjects() {
        return new ArrayList<>(this.sceneObjectsByID.values());
    }

    //	Returns the scene object based on the id
    @Override
    public MSceneObject GetSceneObjectByID(String id) {
        return this.sceneObjectsByID.get(id);
    }

    //	Returns the scene object based on the name
    @Override
    public MSceneObject GetSceneObjectByName(String name) {
        return this.GetSceneObjectByID(this.nameIdMappingSceneObjects.get(name).get(0));
    }

    //	Returns the scene object based on range
    @Override
    public List<MSceneObject> GetSceneObjectsInRange(MVector3 position, double range) {
        return this.sceneObjectsByID.values().parallelStream().filter(s -> MVector3Extensions.euclideanDistance(s.Transform.getPosition(), position) <= range).collect(Collectors.toList());
    }

    //	Returns the collider of the scene objects
    @Override
    public List<MCollider> GetColliders() {
        return this.GetSceneObjects().stream().map(MSceneObject::getCollider).collect(Collectors.toList());
    }

    //	Returns the collider of the scene object based on the id
    @Override
    public MCollider GetColliderById(String id) {

        return this.sceneObjectsByID.get(id).Collider;
    }

    //	Returns the collider of the scene objects based on the range
    @Override
    public List<MCollider> GetCollidersInRange(MVector3 position, double range) {
        return this.GetSceneObjectsInRange(position, range).parallelStream().map(MSceneObject::getCollider).collect(Collectors.toList());
    }

    //	Returns the meshes of the scene objects
    @Override
    public List<MMesh> GetMeshes() {
        return this.GetSceneObjects().stream().map(MSceneObject::getMesh).collect(Collectors.toList());
    }

    //	Returns the meshes of the scene object based on the id
    @Override
    public MMesh GetMeshByID(String id) {
        return this.GetSceneObjectByID(id).getMesh();
    }

    //	Returns the transforms of the scene objects
    @Override
    public List<MTransform> GetTransforms() {
        return this.GetSceneObjects().stream().map(MSceneObject::getTransform).collect(Collectors.toList());
    }

    //	Returns the meshes of the scene object based on the id
    @Override
    public MTransform GetTransformByID(String id) {
        return this.GetSceneObjectByID(id).getTransform();
    }

    //	Returns the avatars
    @Override
    public List<MAvatar> GetAvatars() {
        return new ArrayList<>(this.avatarsByID.values());
    }

    //	Returns the avatar based on the id
    @Override
    public MAvatar GetAvatarByID(String id) {
        return this.avatarsByID.get(id);
    }

    //	Returns the avatar based on the name
    @Override
    public MAvatar GetAvatarByName(String name) {
        //return this.avatarsByID.values().stream().filter(a -> a.Name.equals(name)).findFirst().orElse(null);
        return this.avatarsByID.get(this.nameIdMappingAvatars.get(name).get(0));
    }

    //	Returns the avatar based on the range
    @Override
    public List<MAvatar> GetAvatarsInRange(MVector3 position, double distance) {
        List<MAvatar> _result = new ArrayList<>();
        for (MAvatar avatar : this.avatarsByID.values()) {
            MVector3 avatarPosition = new MVector3();
            avatarPosition.setX(avatar.getPostureValues().getPostureData().get(0));
            avatarPosition.setY(avatar.getPostureValues().getPostureData().get(1));
            avatarPosition.setZ(avatar.getPostureValues().getPostureData().get(2));

            if (MVector3Extensions.euclideanDistance(avatarPosition, position) <= distance) {
                _result.add(avatar);
            }
        }
        return _result;
    }

    @Override
    public double GetSimulationTime() {
        return 0;
    }

    // Returns the changes from the previous frame
    @Override
    public MSceneUpdate GetSceneChanges() {
        return this.sceneUpdate;
    }

    // Returns all sceneobjects and avatars as MSceneUpdate
    @Override
    public MSceneUpdate GetFullScene() {
        // List<MSceneObject> sceneObjects = this.GetSceneObjects();
        MSceneUpdate sceneUpdate = new MSceneUpdate();
        sceneUpdate.setAddedSceneObjects(this.GetSceneObjects());
        sceneUpdate.setAddedAvatars(this.GetAvatars());
        return sceneUpdate;
    }

    @Override
    public MNavigationMesh GetNavigationMesh() {

        return null;
    }

    //Applies the scene manipulation on the scene
    // <param name="sceneUpdates">The scene manipulations to be considered</param>
    public MBoolResponse Apply(MSceneUpdate sceneUpdate) {
        MBoolResponse result = new MBoolResponse(true);
        //Increment the frame id
        this.frameID++;
        //Stores the history
        SceneHistory.offer(new Tuple<>(frameID, sceneUpdate));

        //Only allow the max buffer size
        while (SceneHistory.size() > this.HistoryBufferSize)
            this.SceneHistory.poll();

        //Set the scene changes to the input of the present frame
        this.sceneUpdate = sceneUpdate;

        if (sceneUpdate.isSetAddedAvatars())
            this.AddAvatars(result, sceneUpdate.AddedAvatars);

        //Check if there are new scene objects which should be added
        if (sceneUpdate.isSetAddedSceneObjects())
            this.AddSceneObjects(result, sceneUpdate.AddedSceneObjects);


        //Check if there are changed avatars that need to be retransmitted
        if (sceneUpdate.isSetChangedAvatars())
            this.UpdateAvatars(result, sceneUpdate.ChangedAvatars);

        //Check if there are changed sceneObjects that need to be retransmitted
        if (sceneUpdate.isSetChangedSceneObjects())
            this.UpdateSceneObjects(result, sceneUpdate.ChangedSceneObjects);

        //Check if there are avatars that need to be removed
        if (sceneUpdate.isSetRemovedAvatars())
            this.RemoveAvatars(result, sceneUpdate.RemovedAvatars);

        //Check if there are scene objects that need to be removed
        if (sceneUpdate.isSetRemovedSceneObjects())
            this.RemoveSceneObjects(result, sceneUpdate.RemovedSceneObjects);
        return result;
    }

    //	Removes all scene objects from the scene
    //	<param name="sceneObjectIDs">The IDs of the scene objects which should be removed</param>
    private void RemoveSceneObjects(MBoolResponse result, List<String> removedSceneObjects) {
        for (String id : removedSceneObjects) {
            MSceneObject removed = this.sceneObjectsByID.remove(id);
            if (removed != null) {
                this.nameIdMappingSceneObjects.remove(removed.getName());
            } else {
                result.addToLogData("could not remove scene object  with id: " + id + " not found");
                result.setSuccessful(false);
            }
        }
    }

    //	Removes all scene objects from the scene
    //	<param name="avatarIDs">The IDs of the avatars which should be removed</param>
    private void RemoveAvatars(MBoolResponse result, List<String> removedAvatars) {
        for (String id : removedAvatars) {
            MAvatar removed = this.avatarsByID.remove(id);
            if (removed != null) {
                this.nameIdMappingAvatars.remove(removed.getName());
            } else {
                result.addToLogData("could not remove avatar with id: " + id + " not found");
                result.setSuccessful(false);
            }
        }
    }

    //	Updates all scene objects
    //	<param name="avatars>The scene objects which should be updated</param>
    private void UpdateSceneObjects(MBoolResponse result, List<MSceneObjectUpdate> changedSceneObjects) {
        for (MSceneObjectUpdate object : changedSceneObjects) {
            MSceneObject sceneObject = this.sceneObjectsByID.get(object.ID);
            if (sceneObject != null) {
                //Update the transform
                if (object.isSetTransform()) {
                    MTransformUpdate transformUpdate = object.Transform;
                    if (transformUpdate.isSetPosition())
                        sceneObject.Transform.Position = MVector3Extensions.toMVector3(transformUpdate.Position);

                    if (transformUpdate.isSetRotation())
                        sceneObject.Transform.Rotation = MQuaternionExtensions.toMQuaternion(transformUpdate.Rotation);

                    if (transformUpdate.isSetParent())
                        sceneObject.Transform.Parent = transformUpdate.Parent;
                }

                //update the mesh
                if (object.isSetMesh()) {
                    this.sceneObjectsByID.get(object.ID).setMesh(object.getMesh());
                }

                //update the collider
                if (object.isSetCollider()) {
                    this.sceneObjectsByID.get(object.ID).setCollider(object.getCollider());
                }

                //update the physics
                if (object.isSetPhysicsProperties()) {
                    this.sceneObjectsByID.get(object.ID).setPhysicsProperties(object.getPhysicsProperties());
                }
            } else {
                result.addToLogData("Could not update scene object: " + object.ID + " object not found");
                result.setSuccessful(false);
            }
        }

    }

    //	Updates all avatars
    //	<param name="avatars>The avatars which should be updated</param>
    private void UpdateAvatars(MBoolResponse result, List<MAvatarUpdate> changedAvatars) {
        for (MAvatarUpdate avatarUpdate : changedAvatars) {
            MAvatar avatar = this.avatarsByID.get(avatarUpdate.ID);
            if (avatar != null) {
                if (avatarUpdate.isSetDescription())
                    this.avatarsByID.get(avatarUpdate.getID()).Description = avatarUpdate.getDescription();

                if (avatarUpdate.isSetPostureValues())
                    this.avatarsByID.get(avatarUpdate.getID()).PostureValues = avatarUpdate.getPostureValues();


                if (avatarUpdate.isSetSceneObjects()) {
                    this.avatarsByID.get(avatarUpdate.getID()).SceneObjects = avatarUpdate.getSceneObjects();
                }
            } else {
                result.addToLogData("Could not update avatar : " + avatarUpdate.ID + " not found");
                result.setSuccessful(false);
            }
        }
    }

    //	Adds all scene objects to the scene
    //	<param name="sceneObjects>The scene objects which should be added</param>
    private void AddSceneObjects(MBoolResponse result, List<MSceneObject> addedSceneObjects) {
        for (MSceneObject object : addedSceneObjects) {
            if (this.sceneObjectsByID.putIfAbsent(object.ID, object) != null) {

                result.addToLogData("Could not add scene object: " + object.Name + " is already registered");
                result.setSuccessful(false);
                continue;
            }

            if (!this.nameIdMappingSceneObjects.containsKey(object.Name)) {
                this.nameIdMappingSceneObjects.putIfAbsent(object.Name, new ArrayList<>() {
                    {
                        add(object.ID);
                    }
                });
            } else {
                this.nameIdMappingSceneObjects.get(object.Name).add(object.ID);
            }
        }
    }

    //	Adds all avatars to the scene
    //	<param name="avatars">The avatars which should be added</param>
    private void AddAvatars(MBoolResponse result, List<MAvatar> addedAvatars) {
        for (MAvatar avatar : addedAvatars) {
            if (this.avatarsByID.putIfAbsent(avatar.ID, avatar) != null) {

                result.addToLogData("Could not add avatar: " + avatar.Name + " is already registered");
                result.setSuccessful(false);
                continue;
            }
            if (!this.nameIdMappingAvatars.containsKey(avatar.Name)) {
                this.nameIdMappingAvatars.putIfAbsent(avatar.Name, new ArrayList<>() {
                    {
                        add(avatar.ID);
                    }
                });
            } else {
                this.nameIdMappingSceneObjects.get(avatar.Name).add(avatar.ID);
            }
        }
    }
}
